### ISSUE TITLE
`parse` silently removes spaces from record ids in the sequence output but not in the metadata output

### ISSUE THREAD

**Issue by huddlej** [nextstrain/augur#1427]
## Current Behavior

`augur parse` [replaces "forbidden characters" with either underscores or no character at all](https://github.com/nextstrain/augur/blob/7d75fb19941785c5e660f4376afd1db77badc6dd/augur/parse.py#L15) using [str.maketrans](https://docs.python.org/3/library/stdtypes.html#str.maketrans). This approach causes [spaces to be removed from sequence ids during parsing](https://github.com/nextstrain/augur/blob/7d75fb19941785c5e660f4376afd1db77badc6dd/augur/parse.py#L121) but metadata ids are not similarly translated. As a result, any input sequences with spaces in their names produce sequence outputs with no space and metadata outputs with spaces. This mismatch unlinks sequence and metadata records in downstream analyses.

## Expected behavior

Sequence and metadata ids should be identical for each original input record.

## How to reproduce

Create a simple sequence file, parse it, and inspect its outputs.

```bash
# Create a simple FASTA file with a space in the record id.
cat << EOF > example.fasta
>a record|some metadata
ACGT
EOF

# Parse the file.
augur parse \
  --sequences example.fasta \
  --fields strain metadata \
  --output-sequences sequences.fasta \
  --output-metadata metadata.tsv

# Inspect sequences.
cat sequences.fasta
# output is:
# >arecord
# ACGT

# Inspect metadata.
cat metadata.tsv
# output is:
# strain	metadata
# a record	some metadata
```

## Possible solution

The simplest solution to the immediate bug is to apply the same "translate" functionality to the metadata strain id as the sequence id.

This behavior of silently renaming record ids is not ideal, though, and we may also want to consider one of the following:

- provide a command line flag that allows users to opt out of the character translation in all output
- provide a command line argument that allows users to define their own character translation map that applies to both sequences
- alert users when/how many record ids are changed

**Comment by jameshadfield** [comment]
Very similar in spirit to https://github.com/nextstrain/augur/issues/1084. It would be great to fix these bugs as because it's not at all obvious that data's been dropped because of this it's hard to gauge how big a role these bugs have played in peoples' workflows over the years.

**Comment by jameshadfield** [comment]
Ran into this again today, except not with spaces but `)`

Test FASTA:
```
>abc)|USA
ATCG
```

The metadata strain name is "abc)" and the FASTA header is "abc_"

**Comment by victorlin** [comment]
Why modify names at all? This functionality was included in the [initial implementation of `augur parse`](https://github.com/nextstrain/augur/commit/90e47bec024ba20f43eb0753695e878dac2ab45f) without any explicit reasoning. Subsequent modifications to the mapping are either similarly unclear (ffd34bda2c416d3aeeb64a46c7a4ef760ec94239, e39a894026ee75e0bbb158849a1b52df9b445975), removes a mapping (388abaee55e0a58e04bf84125db96857e08f4d66), or is a workaround to match the modification in `augur tree` (88ded123cc2bfaf3633efbbc2f065c52ae73e858).

If modification is for eventual usage in `augur tree`, this issue should be considered together with #1084. The modification should happen in just one of the commands and not both.

**Comment by huddlej** [comment]
> Why modify names at all?

The FASTA format technically* doesn't allow spaces in the record id, since a space is used to delimit the record id and description in the defline like:

```
>record-id record-description
ACTG
```

*Whether this is actually enforced or not depends on the parser, though. Also, some sequence repositories export sequences with spaces in the record ids (e.g., `A/Hong Kong/1/2024` for a flu strain). If we choose to keep spaces in the defline, we'd have accept that a downstream parser might drop part of the real strain name (e.g., producing a record with an id of `A/Hong`). I assume this is why `augur parse` has this opinionated behavior of replacing spaces.

For other characters, I can't think of a good argument for replacing them...

**Comment by genehack** [comment]
> > Why modify names at all?
> 
> The FASTA format technically* doesn't allow spaces in the record id, since a space is used to delimit the record id and description in the defline like:

It also sounds like, from commentary in the linked nextstrain/augur#1084 ([here](https://github.com/nextstrain/augur/issues/1084#issuecomment-1298236067) and [further here](https://github.com/nextstrain/augur/issues/1084#issuecomment-1298547698)) that `iqtree` is also re-writing identifiers, and it's possible that this initial rewriting was an attempt to normalize identifiers upstream of that?

**Comment by jameshadfield** [comment]
> It also sounds like, from commentary in the linked https://github.com/nextstrain/augur/issues/1084 (https://github.com/nextstrain/augur/issues/1084#issuecomment-1298236067 and https://github.com/nextstrain/augur/issues/1084#issuecomment-1298547698) that iqtree is also re-writing identifiers, and it's possible that this initial rewriting was an attempt to normalize identifiers upstream of that?

Yes. Newick trees are the other part of the pipeline where strain names easily cause bugs. From the [spec](https://en.wikipedia.org/wiki/Newick_format), "punctuation characters from the grammar (semicolon, parentheses, comma, and colon) are prohibited [in the name]". Similar to @huddlej's comment above, the actual situation is parser-dependent. E.g. see https://github.com/nextstrain/auspice.us/pull/69 where quotes weren't allowed in the name. 

We should come up with a set of not-allowed characters and use a common function & interface in `augur parse`, `augur tree` and any other function which modifies strain names. I'd propose the default be "error & exit" with an opt in `--replace-with X` option or similar.

In terms of the characters to not allow, I'd add quotes and spaces to the newick list above.

**Comment by huddlej** [comment]
@genehack and I were just chatting about this again and it seems like an ideal long-term solution would be to limit the scope of `augur parse` to _only_ parsing FASTA input into FASTA and metadata outputs and use different `curate`-style commands to transform strain names, prettify values in specific fields, and fix dates. This solution would align more with our switch toward more granular commands for data transformations.

If we had a separate `curate` command for replacing invalid characters in strain names, we could apply that in a way that addresses the constraints for FASTA and Newick specs without trying the transformation to multiple other augur commands.

In the short term, I still think my original simple solution above is the easiest way to fix the current silent bug. I started [a PR with this fix](https://github.com/nextstrain/augur/pull/1712).
